infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java

```diff
@@ -81,29 +76,52 @@ public class DefaultClusterManager implements CorfuReplicationClusterManagerAdap
 
     private ConfigStreamListener configStreamListener;
 
+    @Getter
     private String corfuEndpoint = "localhost:9000";
 
     private DefaultClusterConfig topology;
 
+    @Getter
+    public CorfuReplicationDiscoveryService corfuReplicationDiscoveryService;
+
+    @Getter
+    public TopologyDescriptor topologyConfig;
+
+    public String localEndpoint;
+
+    @Setter
+    private String localNodeId;
+
+    private CountDownLatch waitForTopologyInit = new CountDownLatch(1);
+
     public DefaultClusterManager() {
         topology = new DefaultClusterConfig();
-        localEndpoint = topology.getSourceNodeUuids().get(0);
     }
 
-    @VisibleForTesting
-    public DefaultClusterManager(boolean sinkAsLocalEndpoint) {
-        topology = new DefaultClusterConfig();
-        if (sinkAsLocalEndpoint) {
-            localEndpoint = topology.getSinkNodeUuids().get(0);
-        } else {
-            localEndpoint = topology.getSourceNodeUuids().get(0);
+    @Override
+    public void setLocalEndpoint(String endpoint) {
+        this.localEndpoint = endpoint;
+    }
+
+    @Override
+    public synchronized void updateTopologyConfig(TopologyDescriptor newTopology) {
+        if (newTopology.getTopologyConfigId() > topologyConfig.getTopologyConfigId()) {
+            topologyConfig = newTopology;
+            corfuReplicationDiscoveryService.updateTopology(topologyConfig);
         }
     }
 
+    @Override
+    public void register(CorfuReplicationDiscoveryService corfuReplicationDiscoveryService) {
+        this.corfuReplicationDiscoveryService = corfuReplicationDiscoveryService;
+        localNodeId = corfuReplicationDiscoveryService.getLocalNodeId();
```

Comment:
discovery service gets the local node id from DefaultClusterConfig which is already available here.  Can we fetch it from here itself?

It is also a bit counter-intuitive because ClusterManager gives the topology to discovery service but here it is getting topology-related information from there.

Reply:
this was copied from the deleted CorfuReplicationDiscoveryServiceAdapter.\

True that localNodeId is fetched from the DefaultClusterConfig, but it's kinda dynamic in the sense that the localEndpoint is taken from serverContext. Is there a way to get that endpoint here without hardcoding?

Reply:
Would passing the serverContext.localEndpoint to DefaultClusterManager (via a new constructor) be better?

Reply:
Actually no, I read the classes wrong (mixed `DefaultClusterManager` and `DefaultClusterConfig`). 
We cannot pass the endpoint to DefaultClusterManager for our testing, that would change the way we initiate the plugin.